tool/tsh/kubectl.go

```diff
@@ -0,0 +1,272 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/component-base/cli"
+	"k8s.io/kubectl/pkg/cmd"
+	"k8s.io/kubectl/pkg/cmd/plugin"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/kube/kubeconfig"
+)
+
+var (
+	podForbiddenRe   = regexp.MustCompile(`(?m)Error from server \(Forbidden\): pods "(.*)" is forbidden: User ".*" cannot get resource "pods" in API group "" in the namespace "(.*)"`)
+	clusterForbidden = "[00] access denied"
+	// clusterObjectDiscoveryFailed is printed when kubectl tries to do API discovery
+	// - calling /apis endpoint - but Teleport denies the request. Since it cannot
+	// discover the resources available in the cluster, it prints this message saying
+	// that the cluster does not have pod(s). Since every Kubernetes cluster supports
+	// pods, it's safe to create a resource access request.
+	clusterObjectDiscoveryFailed = regexp.MustCompile(`(?m)the server doesn't have a resource type "pods?"`)
+)
+
+// resourceKind identifies a Kubernetes resource.
+type resourceKind struct {
+	kind            string
+	subResourceName string
+}
+
+// onKubectlCommand re-execs itself if env var `tshKubectlRexec` is not set
+// in order to execute the `kubectl` portion of the code. This is a requirement because
+// `kubectl` calls `os.Exit()` in every code path, and we need to intercept the
+// exit code to validate if the request was denied.
+// When executing `tsh kubectl get pods`, tsh checks if `tshKubectlReexec`. Since
+// it's the user call and the flag is not present, tsh reexecs the same exact
+// the user executed and uses an io.MultiWriter to write the os.Stderr output
+// from the kubectl command into an io.Pipe for analysis. It also sets the env
+// `tshKubectlReexec` in the exec.Cmd.Env and runs the command. When running the
+// command, `tsh` will be recalled, and since `tshKubectlReexec` is set only the
+// kubectl portion of code is executed.
+// On the caller side, once the callee execution finishes, tsh inspects the stderr
+// outputs and decides if creating an access request is appropriate.
+// If the access request is created, tsh waits for the approval and runs the expected
+// command again.
+func onKubectlCommand(cf *CLIConf, args []string) error {
+	if os.Getenv(tshKubectlReexec) == "" {
+		err := runKubectlAndCollectRun(cf, args)
+		return trace.Wrap(err)
+	}
+
+	runKubectlCode(args)
+	return nil
+}
+
+const (
+	// tshKubectlReexec is used to control if tsh should re-exec or execute a kubectl
+	// command.
+	tshKubectlReexec = "tsh_kube_reexec"
```

Comment:
Nit: Since this is an env var:
```suggestion
	tshKubectlReexecEnvVar = "TSH_KUBE_REEXEC"
```